HerPublicWebsite/Controllers/QuestionnaireController.cs

```diff
@@ -135,19 +138,76 @@ public IActionResult Address_Post(AddressViewModel viewModel)
             extraRouteValues: new Dictionary<string, object>
             {
                 { "postcode", viewModel.Postcode },
-                { "number", viewModel.BuildingNameOrNumber }
+                { "buildingNameOrNumber", viewModel.BuildingNameOrNumber }
             }
         );
 
         return RedirectToAction(forwardArgs.Action, forwardArgs.Controller, forwardArgs.Values);
     }
 
-    [HttpGet("address/{postcode}/{number}")]
-    public IActionResult SelectAddress_Get(AddressViewModel viewModel, string postcode, string buildingNameOrNumber)
+    [HttpGet("address/{postcode}/{buildingNameOrNumber}")]
+    public IActionResult SelectAddress_Get(string postcode, string buildingNameOrNumber)
+    {
+
+        //TODO (BEISHER-248): Replace with call to OS Places API to get actual results
+        var viewModel = new SelectAddressViewModel()
+        {
+            Addresses = new List<OsPlacesResult>
+            {
+                new() {
+                    Address = "82 Test Place",
+                    Uprn = "1337"
+                },
+                new ()
+                {
+                    Address = "23 Jellyifsh Place",
+                    Uprn = "420"
+                }
+            }
+        };
+
+        TempData["Addresses"] = JsonSerializer.Serialize(viewModel.Addresses);
+
+        return View("SelectAddress", viewModel);
+    }
+
+    [HttpPost("address/{postcode}/{buildingNameOrNumber}")]
+    public IActionResult SelectAddress_Post(SelectAddressViewModel viewModel, string postcode, string buildingNameOrNumber)
+    {
+        if (!ModelState.IsValid)
+        {
+            return SelectAddress_Get(postcode, buildingNameOrNumber);
+        }
+        
+        try
+        {
+            var addressResults = JsonSerializer.Deserialize<List<OsPlacesResult>>(TempData["Addresses"] as string ?? throw new InvalidOperationException());

```

Comment:
I think this is fine, but are you sure that the ordering of the addresses pre and post serialisation will be the same?